### ISSUE TITLE
24-bit true color support

### ISSUE THREAD

**Issue by sim642** [weechat/weechat#1364]
Throughout the years, 24-bit true color support has been asked about in #weechat but no issue exists about it.

General resources about terminal 24-bit true color support:
* [Original gist with long discussion](https://gist.github.com/XVilka/8346728).
* [Same document in a standalone repository](https://github.com/termstandard/colors/).

## Issues
WeeChat relies on ncurses to handle the terminal and ncurses relies on terminfo to produce the right escapes. According to the true color gist, ncurses 6.1 supposedly supports it but practically it doesn't seem like it, which is likely why there's exactly _zero_ documentation, tutorials or examples about it.

The current possibilities seem to be:
1. **Use a `-direct` `TERM`, e.g. [`xterm-direct`](https://invisible-island.net/xterm/terminfo-contents.html#tic-xterm_direct).**  
   In this terminfo, the 16M colors are simply encoded sequentially into the standard terminal color numbering and terminfo conditionally uses the 24-bit color escape instead. Technically this should allow using them through ncurses just under color numbers including above 255 (I didn't actually test this).

   However, the extended numbering is inconsistent and plain stupid. Increasing terminal color extensions provide the following:
   * `xterm` has: 8 basic colors + 8 bright colors.
   * `xterm-256color` has: 8 basic colors + 8 bright colors + 216 color cube + 24 grayscale colors.
   * `xterm-direct` has: 8 basic colors + 16M true colors.

   While `xterm-256color` is backwards compatible with `xterm` (it keeps all the original colors and just adds new ones), `xterm-direct` is incompatible with both `xterm-256color` _and_ `xterm` because it removes the bright colors, color cube and grayscale colors and replaces them with the beginning of true colors. Technically it works, but it's incredibly stupid on terminfo's part: any application that previously used bright colors or 256 colors extension will just show them as (almost) blacks in the true color scheme. This is so annoying that it's hard to imagine anyone actually wanting to use `xterm-direct` because it makes every terminal program that isn't 24-bit color compatible but uses more than the basic 8 colors unusable and ugly.

   This has also been a major point brought up in the gist discussion. In a sensible world `xterm-direct` would have: 8 basic colors + 8 bright colors + 216 color cube + 24 grayscale colors + 16M true colors. It's yet to be seen if and in how many years this gets improved to make any sense.

2. [**Use terminal palette modification (via `can_change_color` and `init_color`).**](https://gist.github.com/rofl0r/12c766fc7e72f90cb56daf5d15652fc9)  
   Ncurses has even before had the commands to modify RGB values of the 256-color palette in the (possibly rare) cases where it actually is supported. This just means that the existing standard 256-color palette would be replaced with some other colors but that's far from ideal:

   * Redefining a color means that the standard color of that color number isn't usable anymore and would also suddenly change.
   * Still limited by 256 different colors in the palette.
   * WeeChat would have to, in addition to the color pairs table, also manage a colors table to track which colors have been redefined to which RGB values and use them correctly.
   * Changing the palette has nothing to do with the actual 24-bit color escapes even (?).

## WeeChat changes
_If_ there's a solution to the above issues, then WeeChat would still require some changes to make it usable:
- [ ] Allow many more colors in [WeeChat's own color codes](https://weechat.org/files/doc/stable/weechat_dev.en.html#color_codes_in_strings).
- [ ] Allow RGB colors in options (#312).

**Comment by anonymous2ch** [comment]
I'm using latest weechat from git, compiled with ncurses 6.1 20181020, [custom terminfo with setb24 & setf24](https://www.gnu.org/software/emacs/manual/html_node/efaq/Colors-on-a-TTY.html)  commands, launching it with COLORTERM=truecolor and still as soon as weechat launches it only displays 256 colors (even when I run a passthough color test shell script with /exec -sh). I guess there is a converter function somewhere, which converts every displayed color before sending it to ncurses for output). All other terminal apps display 24bit just fine..

**Comment by sim642** [comment]
The `setf24` and `setb24` capabilities are entirely non-standard and unofficial. Standard terminfo doesn't define them and thus ncurses doesn't know anything about them either, so it can't possibly make any use of them and doesn't have any API for them. What `xterm-direct` does is that it defines the standard `setaf` and `setab` capabilities for 16M values instead of just 256.

With `/exec` there is no "passthrough" of output: WeeChat simply parses the ANSI color escapes in the output and uses them when displaying the colors just like any other. So that has the same limitation as WeeChat overall.

**Comment by anonymous2ch** [comment]
It doesn't simply parse ansi codes, but instead converts them to 256 color palette with gui_color_decode_ansi function in https://github.com/weechat/weechat/blob/master/src/gui/gui-color.c

**Comment by sim642** [comment]
It converts them to the 256 color palette because that's all we can properly display using ncurses at this moment. There's no way to just "use the original ANSI codes" because there's no ncurses API to use that as I explained above.

It is not as simple as just writing the 24-bit color ANSI escapes to stdout. WeeChat relies entirely on ncurses to manage the output and efficiently updating the terminal screen when and where needed, e.g. for windows (splits) and window scrolling. Going around ncurses to write escapes to stdout would screw everything up because it wouldn't be in sync with ncurses's internal statekeeping and could get over-rendered at any point by the color specified to ncurses.

Other terminal programs which do support 24-bit color do the whole terminal management from scratch and don't rely on ncurses. That's how they can directly output escape codes. It comes at the price of having to reimplement and maintain everything that ncurses currently offers us.

**Comment by anonymous2ch** [comment]
Well, that's unfortunate, since with xterm-direct my weechat displays only black screen, while consuming 100% CPU (fortunately a single thread of it), cannot even kill it unless with -s 9...

**Comment by sim642** [comment]
WeeChat's current color management has been implemented with at most 256 colors in mind. I had a quick look now what really happens when trying to start under `xterm-direct` and immediately two problem points showed up:

1. WeeChat internally keeps a reverse color pair mapping, which with `COLORS = 16777216` would want to allocate an array of size `(COLORS + 2) * (COLORS + 2) * 2 bytes = 512 terabytes` (or 1024 terabytes with PR #1365): https://github.com/weechat/weechat/blob/d540ea91aef5689d1ee1fa1ce86cad809d225023/src/gui/curses/gui-curses-color.c#L586-L589

2. WeeChat internally also allocates some kind of palette with (some) textual names, which would be a linked list with 16777216 elements: https://github.com/weechat/weechat/blob/d540ea91aef5689d1ee1fa1ce86cad809d225023/src/gui/curses/gui-curses-color.c#L1300-L1319

The second point is actually where WeeChat stays stuck. The first point isn't a problem only because Linux `malloc`-s lazily but would become a problem immediately when the table would actually get used. Obviously these are just more things that would have to get fixed if 24-bit color support ever gets reasonably implemented.

Not working at all under `xterm-direct` currently (without any 24-bit color support) could be considered a separate issue but to me it doesn't seem worth fixing just on its own. Running WeeChat with `xterm-direct` would effectively only give the first 8 terminal colors (followed by 216 near-blacks) so even if it started fine, you'd be better off with `xterm`/`xterm-256color` which have more useful colors.

**Comment by anonymous2ch** [comment]
Hm, in case that's so complicated I'd agree to have a way to map a set of 24 bit colors to a 256 color palette via means of configuration file for starters, since it would actually allow to implement poor's man themes...

**Comment by XVilka** [comment]
Added this bug in my gist and repo 💯

**Comment by diamond-lizard** [comment]
> According to the true color gist, ncurses 6.1 supposedly supports it but practically it doesn't seem like it, which is likely why there's exactly _zero_ documentation, tutorials or examples about it.

On the contrary.  According to [this answer](https://stackoverflow.com/a/63611791) to a StackOverflow question titled "[Does ncurses support TrueColor?](https://stackoverflow.com/questions/63570631/does-ncurses-support-truecolor)", written by Thomas Dickey, ncurses' author:

*"ncurses 6.1 (January 2018) introduced support for direct color, as illustrated in a recap of the ncurses versus slang history. That includes an example program picsmap which uses the RGB extension (documented in [user_caps(5)](https://invisible-island.net/ncurses/man/user_caps.5.html#h3-Recognized-capabilities)).*

*Because the number of colors in 24-bit RGB is larger than the range of numbers supported in the original (signed!) 16-bit numbers (see term(5)), it was necessary to provide for larger numbers. ncurses 6.1 does this with minimal change to existing applications by adding to the opaque TERMINAL structure, and adding functions which can manipulate the extended numbers. It was not necessary to change the ABI (currently 6 since August 2015), because none of the documented features changed their binary interface.*

*To use the RGB feature in ncurses it is necessary to have the proper terminal description. xterm-direct is used for xterm. ... Other terminals have their corresponding flavors. Because the feature is documented, with examples, it is possible for others to customize terminal descriptions as needed.*

*picsmap is one of the test-programs for ncurses, which is available separately as "[ncurses-examples](https://invisible-island.net/ncurses/ncurses-examples.html)". Tutorials are off-topic; the [source-code](https://github.com/ThomasDickey/ncurses-snapshots/blob/master/test/picsmap.c) is readily available."*

**Comment by joshheyse** [comment]
I'm aware that weechat doesn't support truecolor/direct.  Just add some more info, in case this is ever tackled, or to help others who might run across a similar issue.

Weechat crashes when TERM=alacritty-direct.

creating a sh function to use a supported TERM before starting weechat fixes the issue
```
function weechat() {
  TERM=alacritty command weechat
}
```

weechat crash report
```
======= WeeChat backtrace =======
(written by WeeChat 4.1.2, compiled on Dec  4 2023 00:00:00)
001  ??:0 [function ??]
002  ??:0 [function ??]
003  ??:? [function rb_bug_errno]
004  ??:? [function rb_f_kill]
006  ??:0 [function ??]
007  ??:0 [function ??]
008  ??:0 [function ??]
009  ??:0 [function ??]
010  ??:0 [function ??]
011  ??:0 [function ??]
012  ??:0 [function ??]
013  ??:0 [function ??]
014  ??:? [function __libc_start_call_main]
015  :? [function __libc_start_main_alias_2]
016  ??:0 [function ??]
======= End of  backtrace =======
```

LLDB shows the following backtrace on SIGSEGV. 

```
* thread #1, name = 'weechat', stop reason = signal SIGSEGV: invalid permissions for mapped object (fault address: 0xffff7fffc042)
  * frame #0: 0x0000aaab69e97428 weechat`gui_color_get_pair + 328
    frame #1: 0x0000aaab69ea5ec0 weechat`gui_window_set_custom_color_fg.part.0 + 272
    frame #2: 0x0000aaab69ea61c8 weechat`gui_window_string_apply_color_fg + 216
    frame #3: 0x0000aaab69e94770 weechat`gui_bar_window_print_string + 1612
    frame #4: 0x0000aaab69e962c8 weechat`gui_bar_window_draw.part.0 + 3928
    frame #5: 0x0000aaab69e71684 weechat`gui_bar_draw + 224
    frame #6: 0x0000aaab69ea7ee8 weechat`gui_main_refreshes + 200
    frame #7: 0x0000aaab69e13568 weechat`main + 9384
    frame #8: 0x0000ffff851109dc libc.so.6`__libc_start_call_main + 124
    frame #9: 0x0000ffff85110ab0 libc.so.6`__libc_start_main@@GLIBC_2.34 + 156
    frame #10: 0x0000aaab69e13eb0 weechat`_start + 48
```
